resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java

```diff
@@ -145,15 +145,106 @@ public void operationComplete(Future<T> future) throws Exception {
     }
 
     private void internalResolve(Promise<T> promise) {
-        InetSocketAddress nameServerAddrToTry = nameServerAddrs.next();
+        DnsServerAddressStream nameServerAddressStream = getNameServers(hostname);
+
         for (DnsRecordType type: parent.resolveRecordTypes()) {
-            if (!query(hostname, type, nameServerAddrToTry, promise)) {
+            if (!query(hostname, type, nameServerAddressStream, promise)) {
                 return;
             }
         }
     }
 
-    private void query(InetSocketAddress nameServerAddr, final DnsQuestion question, final Promise<T> promise) {
+    /**
+     * Add an authoritative nameserver to the cache if its not a root server.
+     */
+    private void addNameServerToCache(
+            AuthoritativeNameServer name, InetAddress resolved, long ttl) {
+        if (!name.isRootServer()) {
+            // Cache NS record if not for a root server as we should never cache for root servers.
+            parent.authoritativeDnsServerCache().cache(name.domainName(),
+                    additionals, resolved, ttl, parent.ch.eventLoop());
+        }
+    }
+
+    /**
+     * Returns the {@link DnsServerAddressStream} that was cached for the given hostname or {@code null} if non
+     *  could be found.
+     */
+    private DnsServerAddressStream getNameServersFromCache(String hostname) {
+        int len = hostname.length();
+
+        if (len == 0) {
+            // We never cache for root servers.
+            return null;
+        }
+
+        // We always store in the case with a trailing '.'.
+        if (hostname.charAt(len - 1) != '.') {
+            hostname += ".";
+        }
+
+        int idx = hostname.indexOf('.');
+        if (idx <= 0 || idx == hostname.length() - 1) {
+            // We are not interested in handling '.' as we should never serve the root servers from cache.
+            return null;
+        }
+
+        // We start from the closed match and then move down.
+        for (;;) {
+            // Skip '.' as well.
+            hostname = hostname.substring(idx + 1);
+
+            if (hostname.isEmpty()) {
```

Comment:
"hostname.isEmpty()" is never can be, bcz now you are checking "idx/idx2 == hostname.length() - 1".